File: server-go/gphoto_wrapper.go
package main

import (
    "context"
    "encoding/base64"
    "errors"
    "log"
    "os"
    "os/exec"
    "path/filepath"
    "sort"
    "strings"
    "sync"
    "time"
)

type Camera struct {
    Model string `json:"model"`
    Port  string `json:"port"`
}

type GPhotoWrapper struct {
    outputDir        string
    previewDir       string
    isPreviewActive  bool
    previewStopCh    chan struct{}
    mu               sync.Mutex
}

func NewGPhotoWrapper() *GPhotoWrapper {
    gp := &GPhotoWrapper{
        outputDir:     filepath.Join("..", "server", "uploads"),
        previewDir:    filepath.Join("..", "server", "previews"),
        previewStopCh: make(chan struct{}, 1),
    }
    _ = os.MkdirAll(gp.outputDir, 0755)
    _ = os.MkdirAll(gp.previewDir, 0755)
    return gp
}

func (gp *GPhotoWrapper) DetectCamera() ([]Camera, error) {
    cmd := exec.Command("gphoto2", "--auto-detect")
    out, err := cmd.CombinedOutput()
    if err != nil {
        return nil, err
    }
    lines := strings.Split(string(out), "\n")
    cams := make([]Camera, 0)
    for i := 2; i < len(lines); i++ { // skip header lines
        line := strings.TrimSpace(lines[i])
        if line == "" {
            continue
        }
        // parse: <model>  <port> (two or more spaces)
        parts := splitByMultipleSpaces(line)
        if len(parts) >= 2 {
            cams = append(cams, Camera{Model: parts[0], Port: parts[1]})
        }
    }
    return cams, nil
}

func splitByMultipleSpaces(s string) []string {
    // collapse multiple spaces to single and then split
    fields := make([]string, 0)
    cur := strings.Builder{}
    lastSpace := false
    for _, r := range s {
        if r == ' ' || r == '\t' {
            if lastSpace {
                continue
            }
            lastSpace = true
            if cur.Len() > 0 {
                fields = append(fields, cur.String())
                cur.Reset()
            }
        } else {
            lastSpace = false
            cur.WriteRune(r)
        }
    }
    if cur.Len() > 0 {
        fields = append(fields, cur.String())
    }
    return fields
}

func (gp *GPhotoWrapper) CapturePreviewFrame() (map[string]any, error) {
    ts := time.Now().UnixMilli()
    filename := "preview_" + formatInt(ts) + ".jpg"
    path := filepath.Join(gp.previewDir, filename)
    ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
    defer cancel()
    cmd := exec.CommandContext(ctx, "gphoto2", "--capture-preview", "--force-overwrite", "--filename", path)
    out, err := cmd.CombinedOutput()
    if len(out) > 0 {
        msg := string(out)
        if strings.Contains(msg, "Capturing preview frame") {
            log.Printf("GPhoto2 preview message: %s", strings.TrimSpace(msg))
        }
    }
    if err != nil {
        return nil, err
    }
    data, err := os.ReadFile(path)
    if err != nil {
        return nil, errors.New("Preview file not created")
    }
    gp.cleanupPreviews(3)
    b64 := base64.StdEncoding.EncodeToString(data)
    return map[string]any{
        "success":   true,
        "image":     "data:image/jpeg;base64," + b64,
        "timestamp": ts,
    }, nil
}

func (gp *GPhotoWrapper) StartPreviewStream(emit func(event string, payload any), fps int) map[string]any {
    gp.mu.Lock()
    defer gp.mu.Unlock()
    if gp.isPreviewActive {
        return map[string]any{"success": false, "error": "Preview already active"}
    }
    gp.isPreviewActive = true
    interval := time.Second
    if fps > 0 {
        interval = time.Second / time.Duration(fps)
    }
    stopCh := make(chan struct{})
    gp.previewStopCh = stopCh
    go func() {
        ticker := time.NewTicker(interval)
        defer ticker.Stop()
        for {
            select {
            case <-ticker.C:
                res, err := gp.CapturePreviewFrame()
                if err != nil {
                    emit("preview-error", map[string]any{"success": false, "error": err.Error()})
                    continue
                }
                // match client listener name
                emit("previewFrame", res)
            case <-stopCh:
                return
            }
        }
    }()
    return map[string]any{"success": true, "fps": fps}
}

func (gp *GPhotoWrapper) StopPreviewStream() map[string]any {
    gp.mu.Lock()
    defer gp.mu.Unlock()
    if gp.isPreviewActive {
        gp.isPreviewActive = false
        select {
        case gp.previewStopCh <- struct{}{}:
        default:
        }
        gp.previewStopCh = make(chan struct{}, 1)
    }
    return map[string]any{"success": true}
}

func (gp *GPhotoWrapper) CaptureImage() (map[string]any, error) {
    ts := time.Now().UnixMilli()
    filename := "photo_" + formatInt(ts) + ".jpg"
    path := filepath.Join(gp.outputDir, filename)
    ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
    defer cancel()
    cmd := exec.CommandContext(ctx, "gphoto2", "--capture-image-and-download", "--filename", path, "--skip-existing")
    if err := cmd.Run(); err != nil {
        // stderr can contain non-error messages; still return error if command failed
        return nil, err
    }
    // Find created file (case-insensitive extension)
    entries, _ := os.ReadDir(gp.outputDir)
    var matched string
    for _, e := range entries {
        name := e.Name()
        if strings.HasPrefix(name, "photo_"+formatInt(ts)) && (strings.HasSuffix(strings.ToLower(name), ".jpg") || strings.HasSuffix(strings.ToLower(name), ".jpeg") || strings.HasSuffix(strings.ToLower(name), ".jpg")) {
            matched = name
            break
        }
    }
    if matched == "" {
        return nil, errors.New("Photo file not found")
    }
    return map[string]any{
        "success":   true,
        "filepath":  filepath.Join(gp.outputDir, matched),
        "filename":  matched,
        "url":       "/uploads/" + matched,
        "timestamp": ts,
    }, nil
}

func (gp *GPhotoWrapper) cleanupPreviews(keepLast int) {
    entries, err := os.ReadDir(gp.previewDir)
    if err != nil {
        return
    }
    type item struct{ name string; path string; time int64 }
    items := make([]item, 0)
    for _, e := range entries {
        name := e.Name()
        if !strings.HasPrefix(name, "preview_") { continue }
        p := filepath.Join(gp.previewDir, name)
        st, err := os.Stat(p)
        if err != nil { continue }
        items = append(items, item{name: name, path: p, time: st.ModTime().UnixMilli()})
    }
    sort.Slice(items, func(i, j int) bool { return items[i].time > items[j].time })
    for idx := keepLast; idx < len(items); idx++ {
        _ = os.Remove(items[idx].path)
    }
}

func formatInt(v int64) string {
    // fast int64 to string
    if v == 0 { return "0" }
    var b [20]byte
    i := len(b)
    for v > 0 {
        i--
        b[i] = byte('0' + v%10)
        v /= 10
    }
    return string(b[i:])
}

File: server-cpp/src/gphoto_wrapper.cpp
#include "../include/server.h"

GPhotoWrapper::GPhotoWrapper(const std::string& outputDir, const std::string& previewDir)
    : outputDir(outputDir), previewDir(previewDir), isPreviewActive(false) {
    createDirectories(outputDir);
    createDirectories(previewDir);
}

GPhotoWrapper::~GPhotoWrapper() {
    if (isPreviewActive) {
        stopPreviewStream();
    }
}

std::vector<Camera> GPhotoWrapper::detectCamera() {
    std::vector<Camera> cameras;
    std::string result = executeCommand("gphoto2 --auto-detect");
    std::vector<std::string> lines = splitString(result, '\n');
    for (size_t i = 2; i < lines.size(); ++i) {
        std::string line = lines[i];
        if (line.empty()) continue;
        std::vector<std::string> parts;
        std::string current;
        bool inSpace = false;
        for (char c : line) {
            if (c == ' ' || c == '\t') {
                if (!inSpace && !current.empty()) {
                    parts.push_back(current);
                    current.clear();
                }
                inSpace = true;
            } else {
                current += c;
                inSpace = false;
            }
        }
        if (!current.empty()) {
            parts.push_back(current);
        }
        if (parts.size() >= 2) {
            Camera camera;
            camera.model = parts[0];
            camera.port = parts[1];
            cameras.push_back(camera);
        }
    }
    return cameras;
}

std::map<std::string, std::string> GPhotoWrapper::capturePreviewFrame() {
    std::map<std::string, std::string> result;
    std::string timestamp = getCurrentTimestamp();
    std::string filename = "preview_" + timestamp + ".jpg";
    std::string path = previewDir + "/" + filename;
    std::string command = "gphoto2 --capture-preview --force-overwrite --filename " + path;
    std::string cmdResult = executeCommand(command);
    std::vector<unsigned char> imageData = readImageFile(path);
    if (imageData.empty()) {
        result["success"] = "false";
        result["error"] = "Preview file not created";
        return result;
    }
    std::vector<unsigned char> processedData = effects.applyEffect(imageData);
    cleanupPreviews(3);
    std::string base64Image = "data:image/jpeg;base64," + base64Encode(processedData);
    result["success"] = "true";
    result["image"] = base64Image;
    result["timestamp"] = timestamp;
    return result;
}

std::map<std::string, std::string> GPhotoWrapper::startPreviewStream(
    std::function<void(const std::string&, const std::map<std::string, std::string>&)> emit,
    int fps) {
    std::map<std::string, std::string> result;
    if (isPreviewActive) {
        result["success"] = "false";
        result["error"] = "Preview already active";
        return result;
    }
    isPreviewActive = true;
    std::thread previewThread([this, emit, fps]() {
        int interval = 1000 / fps;
        while (isPreviewActive) {
            auto frame = capturePreviewFrame();
            if (frame["success"] == "true") {
                emit("previewFrame", frame);
            } else {
                emit("preview-error", frame);
            }
            std::this_thread::sleep_for(std::chrono::milliseconds(interval));
        }
    });
    previewThread.detach();
    result["success"] = "true";
    result["fps"] = std::to_string(fps);
    return result;
}

std::map<std::string, std::string> GPhotoWrapper::stopPreviewStream() {
    std::map<std::string, std::string> result;
    if (isPreviewActive) {
        isPreviewActive = false;
        result["success"] = "true";
    } else {
        result["success"] = "false";
        result["error"] = "Preview not active";
    }
    return result;
}

std::map<std::string, std::string> GPhotoWrapper::captureImage() {
    std::map<std::string, std::string> result;
    std::string timestamp = getCurrentTimestamp();
    std::string filename = "photo_" + timestamp + ".jpg";
    std::string path = outputDir + "/" + filename;
    std::string command = "gphoto2 --capture-image-and-download --filename " + path + " --skip-existing";
    std::string cmdResult = executeCommand(command);
    std::vector<unsigned char> imageData = readImageFile(path);
    if (imageData.empty()) {
        result["success"] = "false";
        result["error"] = "Photo file not created";
        return result;
    }
    std::vector<unsigned char> processedData = effects.applyEffect(imageData);
    if (!processedData.empty()) {
        writeImageFile(path, processedData);
    }
    result["success"] = "true";
    result["filename"] = filename;
    result["filepath"] = path;
    result["url"] = "/uploads/" + filename;
    result["timestamp"] = timestamp;
    return result;
}

void GPhotoWrapper::cleanupPreviews(int keepLast) {
}

void GPhotoWrapper::setEffect(EffectType effect, const EffectParams& params) {
    effects.setEffect(effect, params);
}

std::pair<EffectType, EffectParams> GPhotoWrapper::getCurrentEffect() const {
    return effects.getEffect();
}

std::string GPhotoWrapper::executeCommand(const std::string& command) {
    std::string result;
    FILE* pipe = popen(command.c_str(), "r");
    if (!pipe) {
        std::cerr << "Error executing command: " << command << std::endl;
        return "";
    }
    char buffer[128];
    while (fgets(buffer, sizeof(buffer), pipe) != nullptr) {
        result += buffer;
    }
    int exitCode = pclose(pipe);
    return result;
}

std::vector<unsigned char> GPhotoWrapper::readImageFile(const std::string& filePath) {
    std::vector<unsigned char> data;
    std::ifstream file(filePath, std::ios::binary);
    if (!file) {
        std::cerr << "Error opening file for reading: " << filePath << std::endl;
        return data;
    }
    file.seekg(0, std::ios::end);
    size_t fileSize = file.tellg();
    file.seekg(0, std::ios::beg);
    data.resize(fileSize);
    if (!file.read(reinterpret_cast<char*>(data.data()), fileSize)) {
        std::cerr << "Error reading file: " << filePath << std::endl;
        data.clear();
    }
    return data;
}

bool GPhotoWrapper::writeImageFile(const std::string& filePath, const std::vector<unsigned char>& data) {
    std::ofstream file(filePath, std::ios::binary);
    if (!file) {
        std::cerr << "Error opening file for writing: " << filePath << std::endl;
        return false;
    }
    if (!file.write(reinterpret_cast<const char*>(data.data()), data.size())) {
        std::cerr << "Error writing file: " << filePath << std::endl;
        return false;
    }
    return true;
}

std::string GPhotoWrapper::base64Encode(const std::vector<unsigned char>& data) {
    const std::string base64Chars = 
        "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
        "abcdefghijklmnopqrstuvwxyz"
        "0123456789+/";
    std::string result;
    int i = 0;
    int j = 0;
    unsigned char charArray3[3];
    unsigned char charArray4[4];
    for (unsigned char c : data) {
        charArray3[i++] = c;
        if (i == 3) {
            charArray4[0] = (charArray3[0] & 0xfc) >> 2;
            charArray4[1] = ((charArray3[0] & 0x03) << 4) + ((charArray3[1] & 0xf0) >> 4);
            charArray4[2] = ((charArray3[1] & 0x0f) << 2) + ((charArray3[2] & 0xc0) >> 6);
            charArray4[3] = charArray3[2] & 0x3f;
            for (i = 0; i < 4; i++) {
                result += base64Chars[charArray4[i]];
            }
            i = 0;
        }
    }
    if (i) {
        for (j = i; j < 3; j++) {
            charArray3[j] = '\0';
        }
        charArray4[0] = (charArray3[0] & 0xfc) >> 2;
        charArray4[1] = ((charArray3[0] & 0x03) << 4) + ((charArray3[1] & 0xf0) >> 4);
        charArray4[2] = ((charArray3[1] & 0x0f) << 2) + ((charArray3[2] & 0xc0) >> 6);
        charArray4[3] = charArray3[2] & 0x3f;
        for (j = 0; j < i + 1; j++) {
            result += base64Chars[charArray4[j]];
        }
        while (i++ < 3) {
            result += '=';
        }
    }
    return result;
}